src/lib.rs

```diff
@@ -317,104 +318,167 @@ impl Renderer {
+                            let pixels = texture.data.as_ptr();
+                            let pixels = pixels.add((width * 4) * y);
+                            std::ptr::copy(pixels, d3d9_memory, width * 4);
+                        }
 
-        let texture_handle = com_ptr_from_fn(|texture_handle| {
-            device.CreateTexture(
-                texture.width,
-                texture.height,
-                1,
-                D3DUSAGE_DYNAMIC,
-                D3DFMT_A8R8G8B8,
-                D3DPOLL_DEFAULT,
-                texture_handle,
-                ptr::null_mut(),
-            )
-        })?;
-
-        let mut locked_rect: D3DLOCKED_RECT = D3DLOCKED_RECT { Pitch: 0, pBits: ptr::null_mut() };
-        hresult(texture_handle.LockRect(0, &mut locked_rect, ptr::null_mut(), 0))?;
-
-        let bits = locked_rect.pBits as *mut u8;
-        let pitch = locked_rect.Pitch as usize;
-        let height = texture.height as usize;
-        let width = texture.width as usize;
-
-        for y in 0..height {
-            let d3d9_memory = bits.add(pitch * y);
-            let pixels = texture.data.as_ptr();
-            let pixels = pixels.add((width * 4) * y);
-            std::ptr::copy(pixels, d3d9_memory, width * 4);
+                        result_texture.UnlockRect(0).unwrap();
+                        fonts.tex_id = TextureId::from(FONT_TEX_ID);
+                        Ok(result_texture)
+                    },
+                    _ => Err(()),
+                }
+            },
+            _ => Err(()),
         }
+    }
+
 
-        texture_handle.UnlockRect(0);
-        fonts.tex_id = TextureId::from(FONT_TEX_ID);
-        Ok(texture_handle)
+    ///IDirect3DDevice9 wrapper
+    #[allow(non_snake_case)]
+    pub unsafe fn Clear(
+        &self,
+        count: u32,
+        prects: *const D3DRECT,
+        flags: u32,
+        color: u32,
+        z: f32,
+        stencil: u32
+    ) -> Result<(), ()>
+    {
+        match self.device.Clear(count ,prects, flags, color, z, stencil){Ok(_) => Ok(()), _ => Err(())}
     }
+
+    ///IDirect3DDevice9 wrapper
+    #[allow(non_snake_case)]
+    pub unsafe fn BeginScene(&self) -> Result<(), ()>
+    {
+        match self.device.BeginScene(){Ok(_) => Ok(()), _ => Err(())}
+    }
+
+    ///IDirect3DDevice9 wrapper
+    #[allow(non_snake_case)]
+    pub unsafe fn EndScene(&self) -> Result<(), ()>
+    {
+        match self.device.EndScene(){Ok(_) => Ok(()), _ => Err(())}
+    }
+
+    ///IDirect3DDevice9 wrapper
+    #[allow(non_snake_case)]
+    pub unsafe fn Present<'a, P0>(
+        &self,
+        psourcerect: *const RECT,
+        pdestrect: *const RECT,
+        hdestwindowoverride: P0,
+        pdirtyregion: *const RGNDATA
+    ) -> Result<(), ()> where
+        P0: windows::core::IntoParam<'a, windows::Win32::Foundation::HWND>,
+    {
+        match self.device.Present(psourcerect, pdestrect, hdestwindowoverride, pdirtyregion){Ok(_) => Ok(()), _ => Err(())}
+    }
+
```

Comment:
Can we remove these? They aren't really relevant to the lib

Reply:
I'm aware, they are needed to make the example work tho, as the ID3DDevice is moved into the renderer struct, and is not shared with the winit's loop. I can make them only enabled when building the example?

Reply:
`IDirect3DDevice9` is `Clone`, so you can clone the device and pass the clone to the renderer. `IDirect3DDevice9` is reference counted so both devices are the same one then.

Reply:
Damn I didn't realize. Fixed.